intra/tunnel.go

```diff
@@ -37,95 +37,85 @@ type Listener interface {
 }
 
 // Tunnel represents an Intra session.
-type Tunnel interface {
-	tunnel.Tunnel
-	// Get the DNSTransport (default: nil).
-	GetDNS() doh.Transport
-	// Set the DNSTransport.  This method must be called before connecting the transport
-	// to the TUN device.  The transport can be changed at any time during operation, but
-	// must not be nil.
-	SetDNS(doh.Transport)
-	// When set to true, Intra will pre-emptively split all HTTPS connections.
-	SetAlwaysSplitHTTPS(bool)
-	// Enable reporting of SNIs that resulted in connection failures, using the
-	// Choir library for privacy-preserving error reports.  `file` is the path
-	// that Choir should use to store its persistent state, `suffix` is the
-	// authoritative domain to which reports will be sent, and `country` is a
-	// two-letter ISO country code for the user's current location.
-	EnableSNIReporter(file, suffix, country string) error
-}
+type Tunnel struct {
+	network.IPDevice
 
-type intratunnel struct {
-	tunnel.Tunnel
-	tcp TCPHandler
-	udp UDPHandler
-	dns doh.Transport
+	sd  *intraStreamDialer
+	pp  *intraPacketProxy
+	sni *tcpSNIReporter
+	tun io.Closer
 }
 
 // NewTunnel creates a connected Intra session.
 //
 // `fakedns` is the DNS server (IP and port) that will be used by apps on the TUN device.
-//    This will normally be a reserved or remote IP address, port 53.
+//
+//	This will normally be a reserved or remote IP address, port 53.
+//
 // `udpdns` and `tcpdns` are the actual location of the DNS server in use.
-//    These will normally be localhost with a high-numbered port.
+//
+//	These will normally be localhost with a high-numbered port.
+//
 // `dohdns` is the initial DOH transport.
-// `tunWriter` is the downstream VPN tunnel.  IntraTunnel.Disconnect() will close `tunWriter`.
-// `dialer` and `config` will be used for all network activity.
 // `listener` will be notified at the completion of every tunneled socket.
-func NewTunnel(fakedns string, dohdns doh.Transport, tunWriter io.WriteCloser, dialer *net.Dialer, config *net.ListenConfig, listener Listener) (Tunnel, error) {
-	if tunWriter == nil {
-		return nil, errors.New("Must provide a valid TUN writer")
-	}
-	core.RegisterOutputFn(tunWriter.Write)
-	t := &intratunnel{
-		Tunnel: tunnel.NewTunnel(tunWriter, core.NewLWIPStack()),
+func NewTunnel(fakedns string, dohdns doh.Transport, tun io.Closer, protector protect.Protector, listener Listener) (t *Tunnel, err error) {
```

Comment:
The IPDevice should be independent of the TUN device.
Remove the `tun` input.

Also, remove the `Disconnect()` method. You are already exposing Close().

In the relay code, you can close the tun device once the IPDevice closes.

Reply:
Yes, I was trying to keep all interface unchanged. `Disconnect()` is exposed to Java as well.

Reply:
It's quite confusing to have both, and it complicates our code. To figure out why we need to pass the tun device is mind-boggling.

It's not a requirement to preserve the old interface, especially for Intra. That requirement only gets in our way.

We need to prioritize not having to worry about the Go backend interface, as that's a major productivity hurdle. That was the point of moving the code to the client repo too (which I understand will happen separately). Perhaps it makes sense to move the code first before future changes. For the Outline integration, let's move the code first (without changes, but proper build), before changing it. I believe that will work better, as it allows use to change the backend and frontend at once.

We can leave the clean up to another PR, but if we want Intra to be a model to copy, we should simplify and clean up the legacy interfaces, so people don't spread bad practices.

Reply:
As we discussed offline, in this PR, let's focus on "refactoring" without changing the old interface. Then I'll copy the code to the Intra repository without any code modification (but with Gradle build script). After that, we can further redesign the interface (e.g. use `IntraDevice`, and only export one single package instead of five).